<img src="http://oci02.img.iteso.mx/identidad_de_instancia_2018/ITESO/Logos%20ITESO/Logo-ITESO-Principal.jpg">

<center><font color=#555555><font size=6> Ingeniería Financiera </font> <br> <br>

<center><font color=#555555><font size=5> ITE1731 - Microestructura y Sistemas de Trading </font> <br> <br> <br>

<center><font color=#555555><font size=5> I.F. Paloma Martinez Gonzalez, Ana Marcela Ramos Gutierrez, Andres Piedras Ayala, Jaime Eduardo Vazquez Guzman </font> <br> <br>
   
<center> <font color= #555555> <font size = 4> Abril 2021 </a> | <font color= #555555> <font size = 4> Repositorio: <a href='https://github.com/'>Link</a></font>

<center> <font color= #555555> <font size = 6> Laboratorio 3 </font> <br> <br> <font color= #555555> <font size = 5> Inversión del Capital </font>

# <font color= #555555> <font size = 5> Abstract </font>

<font color= #6B6B6B> <font size = 2> Este documento fue elaborado por tu **Paloma Martinez, Jaime Eduardo Vazquez, Ana Marcela Ramos y Andres Piedras **, como un trabajo parcial para la materia de **Microestructura y Sistemas de Trading - ITE1731**, la cual es parte del curriculum de la licenciatura en Ingeniería Financiera, ofertada por la universidad **ITESO**. En el presente trabajo se plantea el analisis de Behavioral Finance de un sistema de trading. 

# <font color= #6B6B6B> <font size = 6> 0. Librerías y Dependencias </font>

<hr style="border:0.02in solid gray"> </hr>

# <font color= #6B6B6B> <font size = 5> 0.1 Librerías </font> <font color= #555555>

Para correr este notebook es necesario tener instaladas y/o en el archivo _requirements.txt_ las siguientes librerías :

- pandas>=1.1.1 
- numpy>=1.19.1
- jupyter>=1.0.0

# <font color= #6B6B6B> <font size = 5> 0.2 Dependencias </font> <font color= #555555>

Para correr este notebook es necesario contar con los siguientes archivos y tener MetaTrader 5 instalado en tu ordenador:

- files/MyST_LAB2_AMRG.xlsx : Documento de Marcela
- files/MyST_LAB2_APA.xlsx : Documento de Andres
- files/MyST_LAB2_JEVG.xlsx : Documento de Jaime
- files/MyST_LAB2_PMG.xlsx : Documento de Paloma


<br>

# <font color= #6B6B6B> <font size = 6> 1. Introducción </font>

<hr style="border:0.02in solid gray"> </hr>

Este proyecto tiene como objetivo elaborar multiples herramientas para medir el desempeño de nuestra actividad de trading que hicimos en el laboratorio 2. A grandes razgos, el laboratorio dos consistía en hacer uno o mas trades diarios dentro de la aplicacion MetaTrader 5, todo se hacia manual y tu decidías en que invertir utilizando estrategias de analisis técnico, chartista o el que desearas. En base al laboratorio explicado anteriormente, este laboratorio tiene como objetivo analizar diferentes medidas de desempeño de los diferentes compañeros que conforman este grupo y ver cual fue la razón por la cuál no se obtuvieron los rendimientos esperados o vice versa, cuál fue una estrategia que funcionó. 



# <font color= #6B6B6B> <font size = 6> 2. Objetivos </font>

<hr style="border:0.02in solid gray"> </hr>

Nuestro objetivo principal fue identificar todas las medidas de desempeño del trading que hicimos durante dos semanas y concluir cuales fueron los trades que mas exito tuvieron, cuales fueron los que menos y en general como podriamos crear un sistema de trading con las mejores medidas. 




# <font color= #6B6B6B> <font size = 6> 3. Datos </font>

<hr style="border:0.02in solid gray"> </hr>

In [1]:
import functions as fn
from data import f_leer_archivo
import visualizations as vn

En esta primera parte descargaremos los historicos de trading dependiendo el histórico del compañero que queramos analizar. Solamente es necesario cambiar las iniciales del alumno en el nombre del documento para descargar cada una de las tablas. 

In [28]:
#Funcion desde regresa un dataframe con toda la informacion del archivo.
data = f_leer_archivo('MyST_LAB2_AMRG.xlsx','Historico MT4')
data.head()

Time   Position  Symbol Type  Volume    Price    S / L  \
0  2021.02.19 08:10:39  873210071  EURUSD  buy    0.10  1.20965  1.20866   
1  2021.02.22 03:52:22  874804645  EURCHF  buy    1.08  1.08747  1.08643   
2  2021.02.23 08:51:46  876896332  EURUSD  buy   10.00  1.21672  1.21565   
3  2021.02.23 21:03:19  878179027  USDCHF  buy   10.00  0.90353      NaN   
4  2021.02.23 21:14:29  878190806  GBPUSD  buy   10.00  1.41096      NaN   

     T / P               Time.1  Price.1  Commission  Swap   Profit  
0  1.21066  2021.02.19 09:17:50  1.21066           0   0.0    10.10  
1  1.08945  2021.02.22 09:56:00  1.08945           0   0.0   237.67  
2  1.21865  2021.02.23 10:36:20  1.21565           0   0.0 -1070.00  
3  0.90368  2021.02.23 21:14:39  0.90368           0   0.0   165.99  
4  1.41283  2021.02.24 01:09:57  1.41138           0  -2.0   420.00

Una vez descargada la tabla, procederemos a convertir las columnas de opentime y closetime a tipo datetime, utilizando la función de pandas to_datetime. Además, en el mismo DataFrame, creamos una nueva columna nombrada 'time', la cual incluye los segundos transcurridos entre el momento de apertura y el momento de cierre de cada operación.

In [3]:
#Funcion donde el mismo Dataframe convertimos las columnas a tipo Datetime
data2 = fn.f_columnas_tiempos(data)
data2.head()

Time   Position  Symbol Type  Volume    Price    S / L  \
0  2021.02.19 08:10:39  873210071  EURUSD  buy    0.10  1.20965  1.20866   
1  2021.02.22 03:52:22  874804645  EURCHF  buy    1.08  1.08747  1.08643   
2  2021.02.23 08:51:46  876896332  EURUSD  buy   10.00  1.21672  1.21565   
3  2021.02.23 21:03:19  878179027  USDCHF  buy   10.00  0.90353      NaN   
4  2021.02.23 21:14:29  878190806  GBPUSD  buy   10.00  1.41096      NaN   

     T / P               Time.1  Price.1  Commission  Swap   Profit  \
0  1.21066  2021.02.19 09:17:50  1.21066           0   0.0    10.10   
1  1.08945  2021.02.22 09:56:00  1.08945           0   0.0   237.67   
2  1.21865  2021.02.23 10:36:20  1.21565           0   0.0 -1070.00   
3  0.90368  2021.02.23 21:14:39  0.90368           0   0.0   165.99   
4  1.41283  2021.02.24 01:09:57  1.41138           0  -2.0   420.00   

            open_time          close_time     time  
0 2021-02-19 08:10:39 2021-02-19 09:17:50   4031.0  
1 2021-02-22 03:52:22 2021-02-22 09:56:00  21818.0  
2 2021-02-23 08:51:46 2021-02-23 10:36:20   6274.0  
3 2021-02-23 21:03:19 2021-02-23 21:14:39    680.0  
4 2021-02-23 21:14:29 2021-02-24 01:09:57  14128.0

En el mismo DataFrame utilizado anteriormente, agregamos agregamos una nueva columna: 'pips', que es la columna de pérdida o ganancia de la operación expresada en pips, también está 'pips_acm' que son los pips acumulados por operación (la suma acumulativa de la columna pips), y finalmente tienes 'profit_acm', que es el profit en capital, que acumula la cuenta.

In [4]:
def f_pip_size(Ticker):
    pip_size = mt5.symbol_info(Ticker)._asdict().get('trade_tick_size')*10
    n = 1/pip_size
    return int(n)


In [5]:
#Funcion para obtener el número multiplicador para expresar la diferencia de precios en pips.
data3 = fn.f_columnas_pips(data2)
data3.head()

Time   Position  Symbol Type  Volume    Price    S / L  \
0  2021.02.19 08:10:39  873210071  EURUSD  buy    0.10  1.20965  1.20866   
1  2021.02.22 03:52:22  874804645  EURCHF  buy    1.08  1.08747  1.08643   
2  2021.02.23 08:51:46  876896332  EURUSD  buy   10.00  1.21672  1.21565   
3  2021.02.23 21:03:19  878179027  USDCHF  buy   10.00  0.90353      NaN   
4  2021.02.23 21:14:29  878190806  GBPUSD  buy   10.00  1.41096      NaN   

     T / P               Time.1  Price.1  Commission  Swap   Profit  \
0  1.21066  2021.02.19 09:17:50  1.21066           0   0.0    10.10   
1  1.08945  2021.02.22 09:56:00  1.08945           0   0.0   237.67   
2  1.21865  2021.02.23 10:36:20  1.21565           0   0.0 -1070.00   
3  0.90368  2021.02.23 21:14:39  0.90368           0   0.0   165.99   
4  1.41283  2021.02.24 01:09:57  1.41138           0  -2.0   420.00   

            open_time          close_time     time  pips  pips_acm  profit_acm  
0 2021-02-19 08:10:39 2021-02-19 09:17:50   4031.0  10.1      10.1       10.10  
1 2021-02-22 03:52:22 2021-02-22 09:56:00  21818.0  19.8      29.9      247.77  
2 2021-02-23 08:51:46 2021-02-23 10:36:20   6274.0 -10.7      19.2     -822.23  
3 2021-02-23 21:03:19 2021-02-23 21:14:39    680.0   1.5      20.7     -656.24  
4 2021-02-23 21:14:29 2021-02-24 01:09:57  14128.0   4.2      24.9     -236.24

Despues, creamos una funcion que se llama **f_estadisticas_ba** donde la salida es un diccionario con dos llaves una con una tabla con medidas de desempeño que lo podríamos llamar un resumen de nuestro trading y una segunda tabla con el ranking de nuestros trades. 

In [6]:
#Funcion donde obtenemos un diccionario con dos dataframes donde nos dice las estadisticas basicas y raking.
data4 = fn.f_estadisticas_ba(data3)
data4

{'df_1_tabla':              medida        valor                            descripcion
 0       Ops totales    15.000000                    Operaciones totales
 1         Ganadoras    10.000000                  Operaciones ganadoras
 2       Ganadoras_c     6.000000        Operaciones ganadoras de compra
 3       Ganadoras_v     4.000000         Operaciones ganadoras de venta
 4        Perdedoras     5.000000                 Operaciones perdedoras
 5      Perdedoras_c     4.000000       Operaciones perdedoras de compra
 6      Perdedoras_v     1.000000        Operaciones perdedoras de venta
 7   Mediana(Profit)  1330.320000       Mediana de profit de operaciones
 8     Mediana(Pips)    70.600000         Mediana de pips de operaciones
 9     r_efectividad     0.666667  Ganadoras Totales/Operaciones Totales
 10     r_proporcion     2.000000   Ganadoras Totales/Perdedoras Totales
 11  r_efectividad_c     0.400000   Ganadoras Compra/Operaciones Totales
 12  r_efectividad_v     0.266667    

En el caso de Marce, podemos ver que el ranking mas alto fue de un 66.67% con EURUSD y si vemos la otra tabla, podemos observar que tuvo 6 ganadoras de compra y 4 ganadoras de venta y su mediana de profit de operaciones fue de 1330.32

# <font color= #6B6B6B> <font size = 6> 4. Procesos </font>


# <font color= #6B6B6B> <font size = 3> 4.1 Metricas de atribucion al desempeño </font>


En la segunda parte nuestro objetivo fue ver como iba evolucionando el capital. Con la funcion de **f_evolucion_capital** podiamos observar los rendimientos diarios y los rendimientos acumulados a lo largo del tiempo. Algo con lo que tuvimos problemas fue intentar eliminar los fines de semana ya que esos dias no habria el mercado por lo tanto no teniamos ningun trade. 

In [7]:


#funcion cuya salida contiene un dataframe con fechas dia a dia que se hizo taiding (Quitando fines de semana)
data5 = fn.f_evolucion_capital(data3)
data5.head()


C:\Users\Andres\anaconda3\lib\site-packages\pandas\core\frame.py:4296: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


timestamp  profit_d  profit_acm_d
0 2021-02-19     10.10     100010.10
1 2021-02-22    237.67     100247.77
2 2021-02-23   -904.01      99343.76
3 2021-02-24    420.00      99763.76
4 2021-02-25   1408.65     101172.41

La ultima parte de las metricas de atribucion fue una funcion llamada **f_estadisticas_mad** donde podemos observar una tabla un poco mas elaborada con metricas como lo es el Radio de Sharpe. El Radio de Sharpe es una metrica bastante poderosa ya que nos ayuda a medir el rendimiento promedio excedente sobre la tasa libre de riesgo por unidad de riesgo incurrido. Por ejemplo, en el caso de Marce, ella tuvo un radio de Sharpe Negativo de -17.91 por lo que podemos concluir que sus rendimientos estaban por debajo de los rendimientos de la tasa libre de riesgo. Un sharpe muy alto en terminos financieros represente un excedente de rendimiento.  Otra cosa que analizamos fue Draw Up y Draw Down que quiere decir la minusvalía/plusvalía máxima que se presentó en la evolución del capital. Lo que buscabamos obtener aqui es la fecha inicial, la fecha final y la magnitud del periodo donde la acumulación de pérdida/ganancia de capital resultó ser la mayor dentro del periodo de tiempo analizado **Por ejemplo, en el caso de Jaime, (Explicar los resultados)**



In [8]:
#Función que regrese un DataFrame con los resultados de cada MAD expresadas en términos diarios   
data6 = fn.f_estadisticas_mad(data5)
data6.head()

[*********************100%***********************]  1 of 1 completed


Metrica                     Valor  \
0                  sharpe                  -17.9161   
1      sharpe_actualizado  0    0.02
dtype: float64   
2  drawdown_capi_Fech_Ini       2021-02-26 00:00:00   
3  drawdown_capi_Fech_Fin       2021-03-01 00:00:00   
4           drawdown_capi                -0.0444972   

                             Descripción  
0                           Sharpe Ratio  
1          Sharpe Ratio Fórmula Ajustada  
2  Fecha inicial del DrawDown de Capital  
3    Fecha final del DrawDown de Capital  
4     Máxima pérdida flotante registrada

Para visualizar los draw up y draw down utilizaremos la siguiente grafica

# <font color= #6B6B6B> <font size = 3> 4.2 Behavioral Finance </font>

In [9]:
data7 = fn.f_be_de(data3)
data7

ocurrencias status_quo aversion_perdida sensibilidad_decreciente
0            6     100.0%             0.0%                       No

# <font color= #6B6B6B> <font size = 6> 5. Conlclusiones </font>

Nuestro cerebro toma atajos mentales para poder emitir juicios y decisiones de forma rápida y así responder a determinados estímulos, problemas y situaciones.
Y de ahí nacen los sesgos cognitivos, que son situaciones en las que estos atajos hacen que nuestro procesamiento mental no sea correcto y nos lleve a una distorsión, juicio inexacto o una interpretación ilógica de la información.

Algunos ejemplos de los efectos negativos de los sesgos cognitivos serían:
* Comprar o vender en el momento equivocado 
* Salirse de una operación muy pronto
* Desviarse del plan de trading
* Cortar las pérdidas muy tarde
* No analizar correctamente la nueva información durante el desarrollo de la operación
* Que el stop loss te saque de una operación por ser demasiado ajustado
* No tomar una operación cuando deberíamos hacerlo
* Confiar en la información de medios de comunicación o redes sociales más que en tu propio análisis

Nos damos cuenta que todos los integrantes del equipo de alguna forma uh otra caemos en alguno estos sesgos en algun punto de hacer traiding.